cirq-core/cirq/protocols/hash_from_pickle_test.py

```diff
@@ -0,0 +1,111 @@
+    "cirq/protocols/json_test_data/sympy.E.json",
+    "cirq/protocols/json_test_data/sympy.Equality.json",
+    "cirq/protocols/json_test_data/sympy.EulerGamma.json",
+    "cirq/protocols/json_test_data/sympy.Float.json",
+    "cirq/protocols/json_test_data/sympy.GreaterThan.json",
+    "cirq/protocols/json_test_data/sympy.Integer.json",
+    "cirq/protocols/json_test_data/sympy.LessThan.json",
+    "cirq/protocols/json_test_data/sympy.Mul.json",
+    "cirq/protocols/json_test_data/sympy.Pow.json",
+    "cirq/protocols/json_test_data/sympy.Rational.json",
+    "cirq/protocols/json_test_data/sympy.StrictGreaterThan.json",
+    "cirq/protocols/json_test_data/sympy.StrictLessThan.json",
+    "cirq/protocols/json_test_data/sympy.Symbol.json",
+    "cirq/protocols/json_test_data/sympy.Unequality.json",
+    "cirq/protocols/json_test_data/sympy.pi.json",
+    # RigettiQCSAspenDevice does not pickle
+    "cirq_rigetti/json_test_data/RigettiQCSAspenDevice.json",
+    # TODO(#6674,pavoljuhas) - fix pickling of ProjectorSum
+    "cirq/protocols/json_test_data/ProjectorSum.json",
+)
+
+
+def _is_included(json_filename: str) -> bool:
+    json_posix_path = pathlib.PurePath(json_filename).as_posix()
+    if any(json_posix_path.endswith(t) for t in _EXCLUDE_JSON_FILES):
+        return False
+    if not os.path.isfile(json_filename):
+        return False
+    return True
+
+
+@pytest.fixture(scope='module')
+def pool() -> Iterator[multiprocessing.pool.Pool]:
+    ctx = multiprocessing.get_context("spawn")
+    with ctx.Pool(1) as pool:
+        yield pool
+
+
+def _read_json(json_filename: str) -> Any:
+    obj = cirq.read_json(json_filename)
+    obj = obj[0] if isinstance(obj, list) else obj
+    # trigger possible caching of the hash value
+    if isinstance(obj, Hashable):
+        _ = hash(obj)
+    return obj
+
+
+def test_exclude_json_files_has_valid_entries() -> None:
+    """Verify _EXCLUDE_JSON_FILES has valid entries."""
+    # do not check rigetti files if not installed
+    skip_rigetti = all(m.name != "cirq_rigetti" for m in MODULE_TEST_SPECS)
+    json_file_validates = lambda f: any(
+        m.test_data_path.joinpath(os.path.basename(f)).is_file() for m in MODULE_TEST_SPECS
+    ) or (skip_rigetti and f.startswith("cirq_rigetti/"))
+    invalid_json_paths = [f for f in _EXCLUDE_JSON_FILES if not json_file_validates(f)]
+    assert invalid_json_paths == []
+
+
+@pytest.mark.parametrize(
+    'json_filename',
+    [
+        f"{abs_path}.json"
+        for m in MODULE_TEST_SPECS
+        for abs_path in m.all_test_data_keys()
+        if _is_included(f"{abs_path}.json")
+    ],
+)
+def test_hash_from_pickle(json_filename: str, pool: multiprocessing.pool.Pool):
+    obj_local = _read_json(json_filename)
+    if not isinstance(obj_local, Hashable):
+        return
+    # check if pickling works in the main process for the sake of debugging
+    obj_copy = pickle.loads(pickle.dumps(obj_local))
+    assert obj_copy == obj_local
```

Comment:
Do you think it is worth checking `obj_copy is not obj_local` to ensure that we are not just retrieving the object from the cache?

Reply:
> Do you think it is worth checking `obj_copy is not obj_local` to ensure that we are not just retrieving the object from the cache?

Ack.  Such check would fail.  For example, LineQubit specializes its `__new__` method to make LineQubit(0) singleton; this works also for pickle deserialization.

The main purpose of the pickle round-trip is to fail early if obj pickling does not work.
Previously I got such failures later on when getting a result of a multiprocessing call, but that produced a frozen test without any error message.